src/storage/src/source/types.rs

```diff
@@ -144,17 +144,12 @@ impl HealthStatusUpdate {
 /// conversion to Message.
 #[derive(Debug, Clone)]
 pub struct SourceMessage<Key, Value> {
-    /// The time that an external system first observed the message
-    ///
-    /// Milliseconds since the unix epoch
-    pub upstream_time_millis: Option<i64>,
     /// The message key
     pub key: Key,
     /// The message value
     pub value: Value,
-    /// Headers, if the source is configured to pass them along. If it is, but there are none, it
-    /// passes `Some([])`
-    pub headers: Option<Vec<(String, Option<Vec<u8>>)>>,
+    /// Additional metadata columns requested by the user
+    pub metadata: Row,
```

Comment:
```suggestion
    /// Additional metadata columns requested by the user. An empty `Row` indicates no additional metadata.
    pub metadata: Row,
```

I like the use of `Row`, empty ones are free, and we don't need to mess with `Option`'s!